721 CR - To Integrate Supplier Stock

muralikrishnareddy opened the pull request:
For Supplier Stock Integration CR.

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+            err_barcode = []
+            err_sku = []
+            err_supplier = []
+            prod_ids = []
```
This is not used.

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+                        vendor = self._check_vendor(rec[3])
+                    if not vendor: 
+                        err_supplier.append(rec[3])
+                self._raise_error(file_id[1],err_barcode, err_sku, set(err_supplier))
```
Please move the error checking to its own function as this function is already quite long.

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+                    
+                for cnt in range(1,len(data)):
+                    rec = data[cnt]
+                    product = self._check_product(rec[1], rec[0])
```
The csv DictReader would make this more readable as you would be referencing by column name instead of by a number which means nothing.

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+                    if not isinstance(product, list):
+                        product = [product]
+                    if not isinstance(vendor, list):
+                        vendor = [vendor]
```
This is assigned but never used again.

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+                        vendor = [vendor]
+                    product_obj = self.session.pool.get('product.product').browse(self.session.cr, SUPERUSER_ID, product[0])
+                    
+                    qty = float(rec[2])
```
This can be simplified.

```
qty = rec[2]
#% to Exclude Calculation
if vendor_obj.percent_to_exclude:
    exclude_qty = (vendor_obj.percent_to_exclude / 100.0) * qty
    qty = qty - int(exclude_qty)
product_obj.write({'supplier_stock_integration_qty': qty})
```

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+        return ret
+        
+    def _check_product(self, barcode=False, sku=False):
+        cr = pooler.get_db(self.session.cr.dbname).cursor()
```
I dont think we need a new db cursor here, just pass it in.

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+        cr = pooler.get_db(self.session.cr.dbname).cursor()
+        if barcode and sku:
+            prodids = self.session.pool.get('product.product').search(cr, SUPERUSER_ID, [('magento_barcode','=',barcode), ('magento_supplier_sku','=',sku)])
+            if prodids:
```
This should return the product id if 1 is found, but if we have more than 1 it should throw an error as we wont know which is the correct one to update.

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+                        qty = int(round(qty))                    
+                    product_obj.write({'supplier_stock_integration_qty':qty})
+                    
+                    vendor_prod_ids.remove(product_obj.id)
```
This doesn't handle the case where the qty given is 0. In that case the product should not be removed from the list and should be set as out of stock.

It also doesn't handle the case where the product id does not exist in the list. It should raise a more useful error than the ValueError that will be raised if the product is not in the list.

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+                    if not vendor: 
+                        err_supplier.append(rec[3])
+                self._raise_error(file_id[1],err_barcode, err_sku, set(err_supplier))
+                if vendor:
```
No need for this check, we're already checking if we have a vendor

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+                if vendor:
+                    if not isinstance(vendor, list):
+                        vendor = [vendor]
+                    vendor_obj = self.session.pool.get('res.partner').browse(self.session.cr, SUPERUSER_ID, vendor[0])
```
Here you can just use ```vendor``` . the ```_check_vendor``` function returns the id already and line 145 puts it into the list just so you can use the first element. Remove lines 144 and 145.

SPDermot commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -0,0 +1,187 @@
+                    
+                #Out of Stock Products    
+                if vendor_prod_ids and vendor_obj.flg_sku_out_of_stock:
+                    for productobj in self.session.pool.get('product.product').browse(self.session.cr, SUPERUSER_ID, vendor_prod_ids):
```
This is inefficient. You can write the value 0 to all the rows in one go instead of individually.

```
self.session.pool.get('product.product').write(self.session.cr, SUPERUSER_ID, vendor_prod_ids, {'supplier_stock_integration_qty': 0})
```

SPDermot commented on connector_bots_stock_integration/supplier_stock_integration.py:
```diff
@@ -0,0 +1,11 @@
+    
+    _columns = {
+    'percent_to_exclude':fields.integer('Stock % to Exclude', help="stock % has to exclude for the Supplier while importing the Stock"),
+    'flg_sku_out_of_stock':fields.boolean('Is Out of Stock', help="Barcode not found in sheet has to mark as Out of Stock or not")
```
Please use ```flag_skus_out_of_stock```

Also the description here could be better. ```Flag supplier products as out of stock if they are not included in the supplier integration feed. ```

SPDermot commented on connector_bots_stock_integration/supplier_stock_integration.py:
```diff
@@ -0,0 +1,11 @@
+    _inherit = 'res.partner'
+    
+    _columns = {
+    'percent_to_exclude':fields.integer('Stock % to Exclude', help="stock % has to exclude for the Supplier while importing the Stock"),
```
Description ```Percentage of stock to be excluded from the supplier available quantity when importing stock quantities through the supplier feed```

SPDermot commented:
Thanks @muralikrishnareddy , it's looking good it just needs a few fixes.

Please can you remove all the new db cursor lines. We should only have one cursor for this.

Also can you run a pep8 over the code to improve the consistency of it.

muralikrishnareddy commented:
Hi @SPDermot - Thank you for your Comments. I changed the Code as per your Comments and Committed. I had implemented all the changes informed except writing Errors messages also inside the Main Function. The Error Messages function looks clumsy to me, so i maintained them in a separate function for better readability. Review the New Commit Code.

muralikrishnareddy commented on connector_bots_stock_integration/supplier_stock.py:
```diff
@@ -19,145 +19,220 @@
+        """ Processes the supplier stock feed csv
         """
-        self.backend_adapter.get_supplier_stock(new_cr=new_cr)
+        self.backend_adapter.process_stock_feed(filename)
```
Is this process_stock_file(filename) ?
